File: 프로그래머스/3/43162. 네트워크/네트워크.py
# 네트워크

def dfs(i, visited, computers):
    # 현재 노드를 방문 처리
    visited[i] = True
    # 현재 노드와 연결된 다른 노드를 탐색
    for neighbor in range(len(computers[i])):
        # start와 neighbor가 연결되어 있고 neighbor를 방문하지 않았더라면
        if computers[i][neighbor] == 1 and not visited[neighbor]:
            # 이웃인 neighbor 쪽을 방문한다. 
            dfs(neighbor, visited, computers)

def solution(n, computers):
    visited = [False] * n  # 모든 컴퓨터를 방문하지 않은 상태로 초기화
    answer = 0      # 네트워크 개수

    # 모든 컴퓨터를 탐색
    for i in range(n):
        if not visited[i]:  # 방문하지 않은 컴퓨터가 있으면
            dfs(i, visited, computers)  # DFS로 연결된 컴퓨터를 모두 방문 처리
            answer += 1  # 새로운 네트워크 발견 시 증가

    return answer

File: 프로그래머스/3/43162. 네트워크/네트워크.java
// 네트워크
// 모든 요소를 탐색하는 것이 목적이기 때문에 DFS
// 지금 n, 노드 개수를 알 수 있고 아직 탐색되지 않은 노드를 알 수 있다.
// 따라서 computers를 기준으로 그래프를 구현한 다음 번호가 낮은 노드부터
// 깊이 우선 탐색을 진행
public class Solution {
    private static boolean[] visit;
    private static int[][] computer;

    public int solution(int n, int[][] computers){
        int answer = 0;
        computer = computers;
        // 방문 여부를 저장할 배열
        visit = new boolean[n];

        for (int i = 0; i < n; i++) {
            if (!visit[i]) {
                // 아직 방문하지 않은 노드라면 해당 노드를 시작으로 깊이 우선 탐색
                dfs(i);
                answer++;
            }
        }
        return answer;
    }

    private static void dfs(int now) {
        visit[now] = true; // 현재 노드 방문 처리
        for (int i = 0; i < computer[now].length; i++) {
            // 연결되어 있으며, 방문하지 않은 노드라면
            if (computer[now][i] == 1 && !visit[i]) {
                dfs(i); // 해당 노드를 방문하러 이동
            }
        }
    }
}
